File: frontend/src/services/menuService.js
import apiClient from './api';

export const menuService = {
  // Get menu items for a restaurant
  getMenuByRestaurant: async (restaurantId) => {
    try {
      const response = await apiClient.get(`/menus/restaurant/${restaurantId}`);
      return response.data.data || [];
    } catch (error) {
      console.error('Error fetching menu:', error);
      throw error.response?.data || error;
    }
  },

  // Get menu item by ID
  getMenuItemById: async (menuItemId) => {
    try {
      const response = await apiClient.get(`/menus/${menuItemId}`);
      return response.data.data;
    } catch (error) {
      console.error('Error fetching menu item:', error);
      throw error.response?.data || error;
    }
  },

  // Create menu item
  createMenuItem: async (menuData) => {
    try {
      const response = await apiClient.post('/menus', menuData);
      return response.data.data;
    } catch (error) {
      console.error('Error creating menu item:', error);
      throw error.response?.data || error;
    }
  },

  // Update menu item
  updateMenuItem: async (menuItemId, menuData) => {
    try {
      const response = await apiClient.put(`/menus/${menuItemId}`, menuData);
      return response.data.data;
    } catch (error) {
      console.error('Error updating menu item:', error);
      throw error.response?.data || error;
    }
  },

  // Delete menu item
  deleteMenuItem: async (menuItemId) => {
    try {
      const response = await apiClient.delete(`/menus/${menuItemId}`);
      return response.data.success;
    } catch (error) {
      console.error('Error deleting menu item:', error);
      throw error.response?.data || error;
    }
  },

  // Search menu items
  searchMenuItems: async (restaurantId, query) => {
    try {
      const response = await apiClient.get(`/menus/restaurant/${restaurantId}`, {
        params: { search: query },
      });
      return response.data.data || [];
    } catch (error) {
      console.error('Error searching menu items:', error);
      throw error.response?.data || error;
    }
  },

  // Get menu items by category
  getMenuByCategory: async (restaurantId, category) => {
    try {
      const response = await apiClient.get(`/menus/restaurant/${restaurantId}`, {
        params: { category },
      });
      return response.data.data || [];
    } catch (error) {
      console.error('Error fetching menu by category:', error);
      throw error.response?.data || error;
    }
  },
};

export default menuService;

File: backend/src/main/java/com/bitebridge/service/MenuService.java
package com.bitebridge.service;

import com.bitebridge.exception.ResourceNotFoundException;
import com.bitebridge.model.MenuItem;
import com.bitebridge.repository.MenuItemRepository;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;
import java.util.logging.Logger;
import java.util.stream.Collectors;

@Service
public class MenuService {
    private static final Logger logger = Logger.getLogger(MenuService.class.getName());
    private final MenuItemRepository menuItemRepository;

    public MenuService(MenuItemRepository menuItemRepository) {
        this.menuItemRepository = menuItemRepository;
    }

    public MenuItem getMenuItemById(String id) {
        logger.info("Fetching menu item with id: " + id);
        return menuItemRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Menu item not found with id: " + id));
    }

    public List<MenuItem> getAllMenuItems() {
        logger.info("Fetching all menu items");
        return menuItemRepository.findAll();
    }

    public List<MenuItem> getMenuItemsByRestaurant(String restaurantId) {
        logger.info("Fetching menu items for restaurant: " + restaurantId);
        return menuItemRepository.findByRestaurantId(restaurantId);
    }

    public List<MenuItem> getMenuItemsByCategory(String category) {
        logger.info("Fetching menu items by category: " + category);
        return menuItemRepository.findAll().stream()
                .filter(item -> item.getCategory() != null && item.getCategory().equalsIgnoreCase(category))
                .collect(Collectors.toList());
    }

    public MenuItem createMenuItem(MenuItem menuItem) {
        logger.info("Creating new menu item");
        return menuItemRepository.save(menuItem);
    }

    public MenuItem updateMenuItem(String id, MenuItem itemUpdates) {
        logger.info("Updating menu item with id: " + id);
        MenuItem menuItem = menuItemRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Menu item not found with id: " + id));

        if (itemUpdates.getName() != null) {
            menuItem.setName(itemUpdates.getName());
        }
        if (itemUpdates.getDescription() != null) {
            menuItem.setDescription(itemUpdates.getDescription());
        }
        if (itemUpdates.getCategory() != null) {
            menuItem.setCategory(itemUpdates.getCategory());
        }
        if (itemUpdates.getImageUrl() != null) {
            menuItem.setImageUrl(itemUpdates.getImageUrl());
        }
        if (itemUpdates.getPrice() > 0) {
            menuItem.setPrice(itemUpdates.getPrice());
        }
        if (itemUpdates.getDiscountedPrice() > 0) {
            menuItem.setDiscountedPrice(itemUpdates.getDiscountedPrice());
        }
        menuItem.setVegetarian(itemUpdates.isVegetarian());
        menuItem.setVegan(itemUpdates.isVegan());
        menuItem.setAvailable(itemUpdates.isAvailable());
        if (itemUpdates.getCustomizations() != null) {
            menuItem.setCustomizations(itemUpdates.getCustomizations());
        }

        return menuItemRepository.save(menuItem);
    }

    public void deleteMenuItem(String id) {
        logger.info("Deleting menu item with id: " + id);
        if (!menuItemRepository.existsById(id)) {
            throw new ResourceNotFoundException("Menu item not found with id: " + id);
        }
        menuItemRepository.deleteById(id);
    }

    public MenuItem toggleMenuItemStatus(String id) {
        logger.info("Toggling menu item status for id: " + id);
        MenuItem menuItem = menuItemRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Menu item not found with id: " + id));
        menuItem.setAvailable(!menuItem.isAvailable());
        return menuItemRepository.save(menuItem);
    }

    public long getMenuItemCount() {
        logger.info("Getting total menu item count");
        return menuItemRepository.count();
    }
}
